subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/IndexedNormalizedFileSnapshot.java

```diff
@@ -38,4 +38,32 @@ public String getAbsolutePath() {
     public int getIndex() {
         return index;
     }
+
+    @Override
+    protected boolean hasSameNormalizedPathAs(AbstractNormalizedFileSnapshot that) {
+        if (that instanceof IndexedNormalizedFileSnapshot) {
+            IndexedNormalizedFileSnapshot other = (IndexedNormalizedFileSnapshot) that;
+            int myLength = absolutePath.length() - index;
+            int otherLength = other.absolutePath.length() - other.index;
+            if (myLength != otherLength) {
+                return false;
+            }
+            for (int i = myLength - 1; i >= 0; i--) {
+                if (absolutePath.charAt(index + i) != other.absolutePath.charAt(other.index + i)) {
```

Comment:
:x: I may be missing things here, but this looks wrong. If `otherLength<myLength`, this will fail will an out of bounds exception. If `otherLength>myLength`, then the result is wrong because you're not comparing the extra characters in `other`.

Also I think it would be worth trying with `substring(...).equals(...)`, because we may benefit from native optimizations here.

Reply:
We check for otherLength!=myLength above already.

`substring().equals()` is what this is optimizing away.

Reply:
Ok this was my :eyes: , I missed the length check.

Reply:
Btw I was also surprised that there is no efficient `substring`. But both the profiler locally as well as the times on CI show a small improvement. Maybe in 18.3 this won't be necessary? :)